docs/user-defined-type-functions.md

```diff
@@ -0,0 +1,206 @@
+| ------------- | ------------- | ------------- |
+| `__eq(arg: type)` | `boolean` | overrides the == operator to return true if self is syntactically equal to arg |
+| `is(arg: string)` | `boolean` | returns true if self has the same tag as the argument. List of available tags: "nil", "unknown", "never", "any", "boolean", "number", "string", "singleton", "negation", "union", "intersection", "table", "function", "class"  |
+
+| Static Methods | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `getnegation(arg: type)` | `type` | returns an immutable runtime representation of the argument as a negated type; the argument cannot be an instance of a table or a function |
+| `getsingleton(arg: string \| boolean)` | `type` | returns an immutable runtime representation of the argument as a singleton type |
+| `getunion(...: type)` | `type` | returns an immutable runtime representation of an union type of the arguments; requires the arguments size to be at least 2 |
+| `getintersection(...: type)` | `type` | returns an immutable runtime representation of an intersection type of the arguments |
+| `newtable(props: {[type]: type}?, indexer: {key: type, value: type}?, metatable: type?)` | `type` | returns a mutable runtime representation of a table type; the keys of the table's property must be a runtime representation of the singleton types; if provided the metatable parameter, this table becomes a metatable |
+| `newfunction(parameters: {pack: {type}?, variadic: type?}, returns: {pack: {type}?, variadic: type?})` | `type` | returns a mutable runtime representation of a `function` type |
+| `copy(arg: type)` | `type` | returns a deep copy of the argument |
+
+#### Negation
+
+| Instance Methods | Type | Description |
+| ------------- | ------------- | ------------- |
+| `gettype()` | `type` | returns the `type` being negated |
+
+#### Singleton
+
+| Instance Methods | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `getvalue()` | `string \| boolean` | returns the string value of the  singleton |
+
+#### Table
+
+| Instance Methods | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `setprop(key: type, value: type?)` | `nil` | adds / overrides (if same key exists) a key, value pair to the table's properties; if value is nil, removes the key, value pair; if the key does not exist and the value is nil, nothing happens |
+| `getprop(key: type)` | `type?` | returns the value associated with the key from the table's properties; if the key does not exists, returns nil |
+| `getprops()` | `{[type]: type}` | returns a table of the table's properties |
+| `setindexer(key: type, value: type)` | `nil` | sets the table's indexer with key type as the first argument and value type as the second argument |
+| `getindexer()` | `{key: type, value: type}?` | returns the table's indexer as a table; if the indexer does not exist, returns nil |
```

Comment:
Just realized these aren't compatible with the eventual read/write properties that tables will have in the new solver. Indexers also needs read/write flags too.

Reply:
This is true, but there's probably care that needs to be had here because we don't want people to be able to set up entirely disjoint read/write property types right now.